File: src/main/java/com/music/read/MP3ListCell.java
package com.music.read;

import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.geometry.Orientation;
import javafx.geometry.Pos;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.paint.Color;

import java.io.File;
import java.io.IOException;

/**
 * Created by xupanpan on 09/08/2017.
 */
public class MP3ListCell extends ListCell<MP3Info> {
    private HomeView homeView;

    public MP3ListCell(HomeView homeView) {
        this.homeView = homeView;
        emptyProperty().addListener((obs, wasEmpty, isNowEmpty) -> {
            if (isNowEmpty) {
                setContextMenu(null);
            } else {
                setContextMenu(getCM());
            }
        });

    }

    private ContextMenu getCM() {

        ContextMenu contextMenu = new ContextMenu();
        MenuItem playItem = new MenuItem();
        playItem.setText("播放");
        playItem.setOnAction(event -> {
            MP3Info selectedItem = getListView().getSelectionModel().getSelectedItem();
            homeView.playManager.play(selectedItem);
        });
        MenuItem codeItem = new MenuItem();
        codeItem.setText("乱码处理");
        codeItem.setOnAction(event -> {
            MP3Info selectedItem = getListView().getSelectionModel().getSelectedItem();
            homeView.musicParser.executList(selectedItem, true);

        });
        MenuItem deleteItem = new MenuItem();
        deleteItem.setText("删除");
        deleteItem.setOnAction(event -> {
            MP3Info selectedItem = getListView().getSelectionModel().getSelectedItem();
            if (homeView.playManager.mp3Info == selectedItem) {

                int currentPosition = homeView.playManager.currentPosition;
                homeView.list.remove(selectedItem);
                homeView.playManager.play(currentPosition);
            } else {
                homeView.list.remove(selectedItem);
                 homeView.playManager.updateCurrentPosition();
            }
        });
        MenuItem fileItem = new MenuItem();
        fileItem.setText("查看文件");
        fileItem.setOnAction(event -> {
            MP3Info selectedItem = getListView().getSelectionModel().getSelectedItem();
            if (!selectedItem.mp3File.exists()) {
                return;
            }
            File mp3File = selectedItem.mp3File;
            File parentFile = mp3File.getParentFile();

            try {
                java.awt.Desktop.getDesktop().open(new File(parentFile, "/"));
            } catch (IOException e) {
                e.printStackTrace();
            }

        });
        contextMenu.getItems().addAll(playItem, codeItem, deleteItem, fileItem);

        return contextMenu;
    }

    @Override
    protected void updateItem(MP3Info item, boolean empty) {
        if (!empty && item != null) {
            setGraphic(getItemView(item));
        } else {
            setGraphic(null);
        }
        super.updateItem(item, empty);
    }


    public HBox getItemView(final MP3Info info) {

        HBox itemView = new HBox(10);
        itemView.setAlignment(Pos.CENTER_LEFT);

        CheckBox cb = new CheckBox();
        cb.setSelected(info.isChecked);
        cb.selectedProperty().addListener(new ChangeListener<Boolean>() {
            public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
                info.isChecked = newValue;
            }
        });


        Label index = new Label(String.valueOf(getIndex() + 1));
        index.setPrefWidth(25);

        Label name = new Label(info.fileName);
        name.setWrapText(true);
        name.setPrefWidth(200);

        Label title = new Label(info.title);
        title.setWrapText(true);
        title.setPrefWidth(140);
        Label artist = new Label(info.artist);
        artist.setWrapText(true);
        artist.setPrefWidth(80);
        Label album = new Label(info.album);
        album.setWrapText(true);
        album.setPrefWidth(160);
        Label time = new Label(Utils.getMusicTime(info.time));
        time.setWrapText(true);
        time.setPrefWidth(40);

        Separator s1 = new Separator(Orientation.VERTICAL);
        Separator s2 = new Separator(Orientation.VERTICAL);
        Separator s3 = new Separator(Orientation.VERTICAL);
        Separator s4 = new Separator(Orientation.VERTICAL);

        itemView.getChildren().addAll(cb, index, name, s1, title, s2, artist, s3, album, s4, time);

        if (info.isPlaying) {
            setColorPlaying(index);
            setColorPlaying(name);
            setColorPlaying(title);
            setColorPlaying(artist);
            setColorPlaying(album);
            setColorPlaying(time);
        } else {
            setColorDef(index);
            setColorDef(name);
            setColorDef(title);
            setColorDef(artist);
            setColorDef(album);
            setColorDef(time);
        }

        return itemView;

    }

    private void setColorPlaying(Label label) {
        label.setTextFill(Color.LIGHTSKYBLUE);
    }

    private void setColorDef(Label label) {
        label.setTextFill(Color.BLACK);
    }


}

File: src/com/music/read/MP3ListCell.kt
package com.music.read

import com.jfoenix.controls.JFXCheckBox
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Orientation
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.paint.Color

import java.io.File
import java.io.IOException

/**
 * Created by xupanpan on 09/08/2017.
 */
class MP3ListCell(private val homeView: HomeView) : ListCell<MP3Info>() {
    private val hBox = createView()
    private var cb: JFXCheckBox? = null
    private var index: Label? = null
    private var name: Label? = null
    private var title: Label? = null
    private var artist: Label? = null
    private var album: Label? = null
    private var time: Label? = null
    private val contextMenu = cm

    private val listener = ChangeListener<Boolean> { observable, oldValue, newValue ->
        if (newValue) {
            setContextMenu(null)
        } else {
            setContextMenu(contextMenu)
        }
    }

    private val cm: ContextMenu
        get() {
            val contextMenu = ContextMenu()
            val playItem = MenuItem()
            playItem.text = "播放"

            playItem.onAction = EventHandler {
                val selectedItem = listView.selectionModel.selectedItem
                homeView.playManager.play(selectedItem)
            }

            val codeItem = MenuItem()
            codeItem.text = "乱码处理"
            codeItem.onAction = EventHandler {
                val selectedItem = listView.selectionModel.selectedItem
                homeView.musicParser.executList(selectedItem, true)
            }
            val deleteItem = MenuItem()
            deleteItem.text = "删除"
            deleteItem.onAction = EventHandler {
                val selectedItem = listView.selectionModel.selectedItem
                if (selectedItem.isPlaying) {
                    homeView.playManager.closePlay()
                    homeView.setCurrentPlayTitle(Main.APP_NAME)
                }
                DataManager.instans.removeMP3Info(selectedItem)
            }
            val fileItem = MenuItem()
            fileItem.text = "查看文件"
            fileItem.onAction = EventHandler {
                val selectedItem = listView.selectionModel.selectedItem

                val musicFile = selectedItem.musicFile ?: return@EventHandler
                val parentFile = musicFile.parentFile

                try {
                    java.awt.Desktop.getDesktop().open(File(parentFile, "/"))
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            contextMenu.items.addAll(playItem, codeItem, deleteItem, fileItem)

            return contextMenu
        }

    init {
        emptyProperty().addListener(listener)
    }

    override fun updateItem(item: MP3Info?, empty: Boolean) {
        super.updateItem(item, empty)

        if (!empty && item != null) {
            updateView(item)
            graphic = hBox
        } else {
            graphic = null
        }

    }


    fun createView(): HBox {
        val itemView = HBox(10.0)
        itemView.alignment = Pos.CENTER_LEFT
        cb = JFXCheckBox()
        index = Label()
        index!!.prefWidth = 25.0

        name = Label()
        name!!.isWrapText = true
        name!!.prefWidth = 200.0

        title = Label()
        title!!.isWrapText = true
        title!!.prefWidth = 140.0
        artist = Label()
        artist!!.isWrapText = true
        artist!!.prefWidth = 80.0
        album = Label()
        album!!.isWrapText = true
        album!!.prefWidth = 160.0
        time = Label()
        time!!.isWrapText = true
        time!!.prefWidth = 40.0
        val s1 = Separator(Orientation.VERTICAL)
        val s2 = Separator(Orientation.VERTICAL)
        val s3 = Separator(Orientation.VERTICAL)
        val s4 = Separator(Orientation.VERTICAL)
        itemView.children.addAll(cb, index, name, s1, title, s2, artist, s3, album, s4, time)
        return itemView

    }

    private fun updateView(info: MP3Info) {

        cb!!.isSelected = info.isChecked
        cb!!.selectedProperty().addListener { observable, oldValue, newValue -> info.isChecked = newValue!! }
        index!!.text = (getIndex() + 1).toString()
        name!!.text = info.fileName
        title!!.text = info.title
        artist!!.text = info.artist
        album!!.text = info.album
        time!!.text = Utils.getMusicTime(info.time)

        if (info.isPlaying) {
            setColorPlaying(index!!)
            setColorPlaying(name!!)
            setColorPlaying(title!!)
            setColorPlaying(artist!!)
            setColorPlaying(album!!)
            setColorPlaying(time!!)
        } else {
            setColorDef(index!!)
            setColorDef(name!!)
            setColorDef(title!!)
            setColorDef(artist!!)
            setColorDef(album!!)
            setColorDef(time!!)
        }
    }


    private fun setColorPlaying(label: Label) {
        label.textFill = Color.LIGHTSKYBLUE
    }

    private fun setColorDef(label: Label) {
        label.textFill = Color.BLACK
    }


}
